src/mcts/search.cc

```diff
@@ -1007,18 +1080,55 @@ void SearchWorker::ExtendNode(Node* node) {
       if (state != FAIL) {
         // If the colors seem backwards, check the checkmate check above.
         if (wdl == WDL_WIN) {
-          node->MakeTerminal(GameResult::BLACK_WON);
+          result = GameResult::BLACK_WON;
+          trigger = CertaintyTrigger::TB_HIT;
         } else if (wdl == WDL_LOSS) {
-          node->MakeTerminal(GameResult::WHITE_WON);
+          result = GameResult::WHITE_WON;
+          trigger = CertaintyTrigger::TB_HIT;
         } else {  // Cursed wins and blessed losses count as draws.
-          node->MakeTerminal(GameResult::DRAW);
+          result = GameResult::DRAW;
+          trigger = CertaintyTrigger::NORMAL;
         }
         search_->tb_hits_.fetch_add(1, std::memory_order_acq_rel);
         return;
       }
     }
   }
+}
+
+void SearchWorker::ExtendNode(Node* node) {
+  // Initialize position sequence with pre-move position.
+  history_.Trim(search_->played_history_.GetLength());
+  std::vector<Move> to_add;
+  // Could instead reserve one more than the difference between history_.size()
+  // and history_.capacity().
+  to_add.reserve(60);
+  Node* cur = node;
+  while (cur != search_->root_node_) {
+    Node* prev = cur->GetParent();
+    to_add.push_back(prev->GetEdgeToNode(cur)->GetMove());
+    cur = prev;
+  }
+  for (int i = to_add.size() - 1; i >= 0; i--) {
+    history_.Append(to_add[i]);
+  }
 
+  // We don't need the mutex because other threads will see that N=0 and
+  // N-in-flight=1 and will not touch this node.
+  const auto& board = history_.Last().GetBoard();
+  auto legal_moves = board.GenerateLegalMoves();
+  GameResult result = GameResult::UNDECIDED;
+  CertaintyTrigger trigger = CertaintyTrigger::NONE;
+
+  EvalPosition(node, legal_moves, board, result, trigger);
+
+  if (trigger != CertaintyTrigger::NONE) {
+    if (trigger == CertaintyTrigger::TERMINAL)
+      node->MakeTerminal(result);
+    else
+      node->MakeCertain(result, trigger);
+    return;
```

Comment:
This code path here - does not call node->CreateEdges - thus these nodes have no child edges.
As an example of why this should be a problem - a TBhit will flow into this path creating a certain node with no child edges attached.
If the game advances to the point where the tbhit is now root, we need to select which move to perform (in practice hopefully using dtz root probe limiting, but ignore that for sake of argument).  Since the node has visits, its not a candidate for expansion, so we will enumerate its edges trying to find a child to expand - but there are no edges to enumerate.  Thus search will fail to produce a best move.

Reply:
As far as i see it the HasChildren() check in picknodetoextend does not use n_ but uses the number of edges. 

```
if (node->IsCertain()) {
      return NodeToProcess::TerminalHit(node, depth, 1);
    } else if (!node->HasChildren()) {
      return NodeToProcess::Extension(node, depth);
    }
```

So if no edges are created the node will be a candidate for expansion. The code should work in this case as n_ is not relevant. I ran into no problems in 50.000 games with this, but maybe i am missing something else here. I could just clear n_ or trim but this would lose information behind the node made certain that would have to recomputed. In time trouble that can be problematic. If you pinpoint the problem you see, we could fix this without losing information from previous searches.

Reply:
Okay, I can see why it may manage not to crash now.

But I do think there is an off by one error - the code to reset n_ will reset it to 1 because it sets it to 1+child_sum if greater than 1.  But an unexpanded node should have n ==0.  Thus its n will start from 1. Its q values will be wrong - which will effect fpu calculations.

Also the collision code will get confused because N > 0 even though its unexpanded.  So multiple threads could be scheduled to expand the node simultaneously, causing serious race conditions.  (I'm actually surprised that this didn't give you significant crashes.)

Reply:
Hmm, i can't really remember why i had that >1 condition in the first place. I can see what is troubling you, the case arises only if:
(1) the node is a tb_hit or two-fold and,
(2) it only gets one visit and,
(3) it then gets root.
In this case TryScoreUpdate is true for all threads and the node gets potentially added to eachs threads nodestoprocess for expansion. [The q value is less of an issue, as the prior certain result is just weighted higher which in case of a certain root might actually be desired]. I am a little surprised myself it never showed, especially since i generally test with more threads (PR487) and Ipmanchess's sudden death games have lots of only root expansions and playing on policy.

I think (not currently looking at the code) that i included the > 1 condition in a time prior to tablebases and always wondered what would happen if a terminal with 1 node becomes root. I probably did not want to mess with leelas behaviour in that case. But the issue you raised would have had applied there too. 

So i think one can safely drop the > 1 condition and just ReComputeNfromChildren  in all cases. Then nodes without children would have n_=0, including real terminals, tb hits and two folds while fully preserving the search tree behind a certain node that was established through backpropagation.

Reply:
Your current recomputation starts with 1 then adds child visits - so even without the > 1 condition - its still going to end up with 1.  You will need to check if the node has edges, and if not set n to 0 rather than 1.

Reply:
Ok, i remember now. I wanted the function ReComputeNfromChildren to be generally applicable to all nodes. For nodes with one visit and no children as well as for nodes with children. This does not help for the abovementioned certain nodes (tb hits or two-folds) without children where we want to set n_ to 0 instead of the real number of visits. A solution might be to:

RecomputeNfromChildren for all certain nodes with children and set n_ to 0 for all certain nodes without children.

I think that should cover all cases and preserve the search tree behind a certain node that was established though backpropagation.

Reply:
Code changed so n_=0 for certain nodes without children.

Also changed for self-play (make move). Wouldn't this also make it possible to use TBs in self-play?